File: backend/api/internal/domain/task/task.go
package task

import (
	"errors"
	"strings"
	"time"
)

var ErrEmptyTitle = errors.New("task title must not be empty")

type ID string

type Task struct {
	ID        ID
	Title     string
	Completed bool
	CreatedAt time.Time
}

func New(id ID, title string, createdAt time.Time) (Task, error) {
	normalizedTitle := strings.TrimSpace(title)
	if normalizedTitle == "" {
		return Task{}, ErrEmptyTitle
	}

	return Task{
		ID:        id,
		Title:     normalizedTitle,
		Completed: false,
		CreatedAt: createdAt,
	}, nil
}

File: app/apps/clean-tasks/_domain/task.ts
export type Task = Readonly<{
  id: string;
  title: string;
  completed: boolean;
  createdAt: Date;
}>;

export class EmptyTaskTitleError extends Error {
  constructor() {
    super("task title must not be empty");
    this.name = "EmptyTaskTitleError";
  }
}

type CreateTaskInput = {
  id: string;
  title: string;
  createdAt: Date;
};

export function createTask({ id, title, createdAt }: CreateTaskInput): Task {
  const normalizedTitle = title.trim();
  if (!normalizedTitle) {
    throw new EmptyTaskTitleError();
  }

  return {
    id,
    title: normalizedTitle,
    completed: false,
    createdAt,
  };
}
